Adds the tiny mutation

c420-o opened the pull request:
## About The Pull Request
This adds a new mutation in the game called tiny.  This makes you the size of an action figure 

## What makes this different from dwarfism?
Being tiny allows you to vent, go under tables and grilles, and my personal favorite, be able to be held by other players.
## This sounds super unbalanced.
Sure, until you consider the fact it makes you slow and you take twice the amount of damage. so while you _could_ hypothetically ventcrawl to the Head of Personnel's office and get your all access ID, you would also be slow as hell going back to the vent, and by that time you'd be notice by command, and almost certainly beaten to death.

or worse. _much, much worse._
![image](https://user-images.githubusercontent.com/76887567/106399188-26876480-63e5-11eb-9234-eb06b5a57c8b.png)
## Why It's Good For The Game
It's a new way to torture players while at the same time letting other players mess around. There was a round on Sybil where an admin made everyone tiny and let me tell you, it was extremely fun. This should've been added years ago.
<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog
:cl:
add: Added the tiny mutation
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

c420-o committed:
tinymutation

adds code to the tiny mutation when applicaple

c420-o committed:
Merge remote-tracking branch 'origin/master' into tinymutation

tralezab commented:
We already have tinyness from the shrink ray. Why don't you make this a permanent version of that and add the stuff you did for tiny into that

c420-o commented:
> We already have tinyness from the shrink ray. Why don't you make this a permanent version of that and add the stuff you did for tiny into that

The shrink ray code is made to be temporary, this is meant to be permanent

Qustinnus commented:
what he's saying is that this is just the same code except for the things you added while it should be using the same code for both by using a component or something like that so its not so copy pasta

Qustinnus commented:
on another note I feel like this will break a ton of shit so if you are a main and are reading this please testmerge this first. I can only take one stomach-quality pr a year

Mickyan commented:
give it the squashable element

c420-o commented:
> what he's saying is that this is just the same code except for the things you added while it should be using the same code for both by using a component or something like that so its not so copy pasta
I tried but I lack the skills for that. 
> on another note I feel like this will break a ton of shit so if you are a main and are reading this please testmerge this first. I can only take one stomach-quality pr a year
You'd be surprised. I found one issue
>give it the squashable element
again, out of my element (no pun intented). I'll try and ask the guy who made the squashable element but I dunno if he'd be on board

notfrying1pans commented:
can you drop people in your mouth and eat them in one bite

ATH1909 commented on code/datums/components/tackle.dm:
```diff
@@ -299,6 +301,8 @@
 		attack_mod -= 2
 	if(HAS_TRAIT(sacker, TRAIT_DWARF))
 		attack_mod -= 2
+	if(HAS_TRAIT(target, TRAIT_TINY))
```
```suggestion
	if(HAS_TRAIT(sacker, TRAIT_TINY))
```

ATH1909 replied:
also, why is the tackling attack penalty for being tiny less than the tackling defence penalty for being tiny?

c420-o replied:
Mistake on my part. changed so they're both 4

ATH1909 commented on code/modules/mob/living/carbon/carbon.dm:
```diff
@@ -174,6 +174,8 @@
 		if(HAS_TRAIT(src, TRAIT_DWARF))
 			power_throw--
+		if(HAS_TRAIT(src, TRAIT_TINY))
+			power_throw--
```
```suggestion
			power_throw -= 2
```

ATH1909 commented on code/datums/mutations/body.dm:
```diff
@@ -514,3 +514,39 @@
+	difficulty = 16
+	instability = 5
+	conflicts = list(GIGANTISM, DWARFISM)
+	locked = TRUE
```
do you, uh, have a mutation combination recipe for this yet?

ATH1909 replied:
or is it supposed to be admin-only?

c420-o replied:
> do you, uh, have a mutation combination recipe for this yet?

I don't think it needs one. what would it even be?

necromanceranne replied:
Dwarfism and some other compoment mutation, such as maybe Void Magnet or Spatial Instability.

ATH1909 commented:
~~fetish code smh~~

Sparkezel commented:
~~vore~~

notfrying1pans commented:
new antag gimmick: stack people in boxes and then shove them into a storage implant

maxymax13 commented:
filling a box with tiny little assistants and then throwing them into bridge

c420-o commented:
> can you drop people in your mouth and eat them in one bite

> ~fetish code smh~

> ~vore~

gross



> new antag gimmick: stack people in boxes and then shove them into a storage implant

lmao. atm you can only shove tiny people in pet carriers. dunno if I'd wanna go further than that for balance reasons

c420-o committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into tinymutation

zxaber commented:
Can you shove more than one person into a  pet carrier?

Because turning pet carriers into "assistant cluster grenades" sounds pretty dope. Would be a pretty funny rev gimmick if you could make people go along with it.

c420-o commented:
> Can you shove more than one person into a pet carrier?
> 
> Because turning pet carriers into "assistant cluster grenades" sounds pretty dope. Would be a pretty funny rev gimmick if you could make people go along with it.

I'll have to test that out but it would definitely be awesome

itseasytosee commented:
This shouldn't be a mutation. I don't know what else to make it except for the abductor ray, but this shouldn't be a mutation.
Also fetish content speed merge.

interestingusernam3 commented:
>new antag gimmick: stack people in boxes and then shove them into a storage implant

Can this be considered vore? Discuss.

c420-o commented:
> Can you shove more than one person into a pet carrier?
> 
> Because turning pet carriers into "assistant cluster grenades" sounds pretty dope. Would be a pretty funny rev gimmick if you could make people go along with it.
yes. yes you can.

https://user-images.githubusercontent.com/76887567/106401390-d531a200-63f1-11eb-976b-f648da6c17b9.mp4

c420-o committed:
changes a throw value

added via suggestion by ATH1909

jwc1015 commented:
This is just an idea, but maybe, since it makes you so small, make the overdose threshold of chems be lower, or something? Not sure if it's possible, but having chems affect you differently since you're so small would make sense.

maxymax13 commented:
i present a new antag idea: ant revs. the head rev is normal size but everyone he converts turns into a small man. together they must team up to carry the heads of staff into their anthill

c420-o commented:
> This is just an idea, but maybe, since it makes you so small, make the overdose threshold of chems be lower, or something? Not sure if it's possible, but having chems affect you differently since you're so small would make sense.

good idea. I'll add it after I get squashed working

optimumtact commented:
you're way out of your league kid

c420-o commented:
> you're way out of your league kid

but it works as intented

itseasytosee commented:
>but it works as intented

But it doesn't work as required.

c420-o commented:
> > but it works as intented
> 
> But it doesn't work as required.

dunno what you mean by that

D4C-420 commented:
I played that round on sybil as sec and it was fucking impossible to hit anyone. 2x damage and a slow down means jack shit when clicking on someone with a hit box that small is damn near impossible. This gene with some meth and or ephedrine would be absolute cancer to fight. Gene does need more variety in useful or decent genes but this is not of em. 

> do you, uh, have a mutation combination recipe for this yet?
Also might I suggest this be radioactive + dwarfism

c420-o commented:
> I played that round on sybil as sec and it was fucking impossible to hit anyone. 2x damage and a slow down means jack shit when clicking on someone with a hit box that small is damn near impossible. This gene with some meth and or ephedrine would be absolute cancer to fight. Gene does need more variety in useful or decent genes but this is not of em.

well they're pretty slow. as in, crit slow.
in regards to the meth, I plan on using jwc1015's idea to lower the overdose threshold

> > do you, uh, have a mutation combination recipe for this yet?
> > Also might I suggest this be radioactive + dwarfism

c420-o committed:
Adds squashing

c420-o commented:
Just added squashing

itseasytosee commented:
>Just added squashing

https://en.wikipedia.org/wiki/Crush_fetish

c420-o commented:
> > Just added squashing
> 
> https://en.wikipedia.org/wiki/Crush_fetish

why do you have to make it weird
![image](https://user-images.githubusercontent.com/76887567/106407021-f6ea5380-6408-11eb-9ebe-a6803fc8a5a0.png)

oklews commented:
> can you drop people in your mouth and eat them in one bite

Fetish content moment

D4C-420 commented:
> well they're pretty slow. as in, crit slow.

again even with loser ods on chems people are going to speed this up back or past regular speed and even if they are that slow it would ve VERY hard to fight somone that small without the use of aoe effects

c420-o commented:
> > well they're pretty slow. as in, crit slow.
> 
> again even with loser ods on chems people are going to speed this up back or past regular speed and even if they are that slow it would ve VERY hard to fight somone that small without the use of aoe effects

if you can't fight a weak, tiny player that doesn't mean it's unfair that just means you're unrobust

Fikou commented:
we literally have a shrunken component already why would you do this like that dumbass

Fikou commented on code/datums/mutations/body.dm:
```diff
@@ -514,3 +514,39 @@
+	owner.update_transform()
+	owner.can_be_held = TRUE
+	owner.mob_size = MOB_SIZE_TINY
+	owner.ventcrawler = TRUE
```
this is not a bool??

Fikou commented on code/datums/mutations/body.dm:
```diff
@@ -514,3 +514,39 @@
+	owner.update_transform()
+	owner.can_be_held = FALSE
+	owner.mob_size = MOB_SIZE_HUMAN
+	owner.ventcrawler = FALSE
```
again, not a boolean, and will fuck up monkeys and plastic golems if they manage to get the mutation

Fikou commented on code/datums/mutations/body.dm:
```diff
@@ -514,3 +514,39 @@
+	owner.can_be_held = TRUE
+	owner.mob_size = MOB_SIZE_TINY
+	owner.ventcrawler = TRUE
+	owner.pass_flags = PASSTABLE | PASSGRILLE | PASSMOB
```
you have never reset those pass flags

Fikou commented on code/datums/mutations/body.dm:
```diff
@@ -514,3 +514,39 @@
+	if(..())
+		return
+	ADD_TRAIT(owner, TRAIT_TINY, GENETIC_MUTATION)
+	owner.resize = 0.5
```
dont use resize, use matrix

c420-o replied:
done, will be added in the next commit

Fikou commented on code/datums/mutations/body.dm:
```diff
@@ -514,3 +514,39 @@
+	if(..())
+		return
+	REMOVE_TRAIT(owner, TRAIT_TINY, GENETIC_MUTATION)
+	owner.resize = 2
```
same here

Fikou commented on code/datums/mutations/body.dm:
```diff
@@ -514,3 +514,39 @@
+	owner.ventcrawler = TRUE
+	owner.pass_flags = PASSTABLE | PASSGRILLE | PASSMOB
+	owner.physiology.damage_resistance -= 100
+	owner.add_movespeed_modifier(/datum/movespeed_modifier/shrink_ray)
```
lol

Fikou commented:
> Just added squashing

you didnt commit

LemonInTheDark commented:
This seems like it will end poorly

Tlaltecuhtli commented:
can they shoot gun?

GunHog commented:
Can mini-men pick up and use items? I feel that should have some kind of restriction. After all, how are you supposed to grab something many times your own size?

D4C-420 commented:
> if you can't fight a weak, tiny player that doesn't mean it's unfair that just means you're unrobust

No because for the same reason no one likes pixel hunting this mutation is simply for combat a fucking pain to fight.  People will get around the slow down the same way golems do. Scooters. A man this tiny on a scooter would be damn near impossible to hit without breaking the scooter first. You wont have gene using this to hide in bushes your going to have tiny men on scooters either A shooting guns at you with no risk at being hit due to their hit box or B tiny hulks (still on scooters) beating up assistants who can't fight back because they are playing a damn pixel hunting adventure game.  If I'm fighting against a weak player and lose then I'm unrobust and did not prepare enough. If I'm fighting against a scientist or sec officer with a riot shield and body armor or anomaly armor on a scooter with a hit box size that puts the jank on the megafauna hit box to shame then I'm not power gamming hard enough anymore.

Also watch players use this to make tiny human projectiles out of humanified monkeys.

maxymax13 commented:
solution: just let them not operate a vehicle because that makes no sense, and make guns throw them across the room directly into a wall putting them in crit

also it would be funny if a tiny player with a pen or energy dagger dealt more damage than a regular person with one. pocket traitor

c420-o commented:
> solution: just let them not operate a vehicle because that makes no sense, and make guns throw them across the room directly into a wall putting them in crit

Good Idea

> also it would be funny if a tiny player with a pen was able to deal actual significant damage and one with an energy dagger was able to wield it like a normal esword and reflect projectiles
> ![small man](https://user-images.githubusercontent.com/38844529/106494529-4ed09b00-6499-11eb-9913-4b74a97a4a75.png)

If I can add a way to make inhand icons bigger, sure

c420-o commented:
> Can mini-men pick up and use items? I feel that should have some kind of restriction. After all, how are you supposed to grab something many times your own ~size~ mass?

mass can't be created or destroyed. The in-game explanation is just that the tiny mutation makes it so the cells in your body are bunched up, so you'd still have the same mass and weight, relatively speaking. This is just going off of google and high school, so I could be wrong.

c420-o commented:
>your going to have tiny men on scooters either A shooting guns at you with no risk at being hit due to their hit box or B tiny hulks (still on scooters) beating up assistants who can't fight back because they are playing a damn pixel hunting adventure game.
> 
I'm sorry but this is fucking hilarious to try to visalize. But to answer your question, Honestly, We'll have to wait and see. If they're too hard to hit I'll make it so you're not as small. But for now, I'll leave it the way it is.

Qustinnus commented:
adding more OP mutations is cringe

maxymax13 commented:
it funy tho :(

c420-o commented:
> adding more OP mutations is cringe

any suggestions to make it so it's a negative mutation?

c420-o commented:
misclick whoops

maxymax13 commented:
increase the damage multiplier and make them really easy to flick around. i know they weigh the same because of cell compacting or whatever but gameplay > making something nonsensical slightly less so. a very small human would freeze to death immediately, anyways

c420-o commented:
> increase the damage multiplier and make them really easy to flick around. i know they weigh the same because of cell compacting or whatever but gameplay > making something nonsensical slightly less so. a very small human would freeze to death immediately, anyways. this is a game where potassium reacts extremely violently and vacuum conducts heat and the depressurizing of human causes a death in seconds rather than instant spagehettification

> increase the damage multiplier and make them really easy to flick around
already added.

>very small human would freeze to death immediately, anyways

yeah, I'll add that in the commit, too.

c420-o committed:
adds various things to tiny mutation

people with the tiny mutation can't use guns, and body.dm code improved thanks to the help of floyd and fikou. Thanks guys!

nullAdept commented:
Can you be a tiny hulk? If you can you should probably make it so you can't because that makes no sense.

c420-o commented:
> Can you be a tiny hulk? If you can you should probably make it so you can't because that makes no sense. 

I'll have to experiment with it. Given the damage multiplier and cold resistance multiplier (not added yet) it would probably be more impractical than anything else

c420-o commented:
Thank you all for your feedback and suggestions. You've all been extremely helpful!

MrMelbert commented on code/modules/mob/living/carbon/human/human_helpers.dm:
```diff
@@ -108,6 +108,10 @@
 	if(HAS_TRAIT(src, TRAIT_NOGUNS))
 		to_chat(src, "<span class='warning'>You can't bring yourself to use a ranged weapon!</span>")
 		return FALSE
+	if(G.trigger_guard == TRIGGER_GUARD_NORMAL)
+		if(HAS_TRAIT(src, TRAIT_TINY))
+			to_chat(src, "<span class='warning'>You're too small to reach the trigger!'</span>")
+			return FALSE
```
Trigger guard was checked for `TRIGGER_GUARD_NORMAL` 3 lines prior, you can piggyback on that instead of checking it a second time

c420-o replied:
how'd I get the message though?

c420-o committed:
tiny triggerguard piggybacks off of TRIGGER_GUARD_NORMAL

thanks MrMelbert!

necromanceranne commented:
Smol

moocowswag commented:
Maximum nanite storage should probably be reduced a ton too, if someone has access to the tiny gene they probably have access to nanites.

But I do think it would be fun to build a nanite lab that only tiny people can get into by ventcrawling.

c420-o commented:
atm this is stalled until I figure out a way to test metabolism

Megamaw commented:
suggestion if this ends up being too strong:

if you have a flyswatter and click on a tile that has someone with this mutation on it, regardless if you clicked on their sprite, it should crit them (or deal a shitload)

thematically appropriate, an item that afaik a lot of departments don't have ready access to, gives niche counterplay if this ends up being completely impossible to combat (which would surprise me.)

moocowswag commented:
> if you have a flyswatter and click on a tile that has someone with this mutation on it, regardless if you clicked on their sprite, it should crit them (or deal a shitload)

Insta gib fly people with the tiny mutation?

c420-o commented:
> suggestion if this ends up being too strong:
> 
> if you have a flyswatter and click on a tile that has someone with this mutation on it, regardless if you clicked on their sprite, it should crit them (or deal a shitload)
> 
> thematically appropriate, an item that afaik a lot of departments don't have ready access to, gives niche counterplay if this ends up being completely impossible to combat (which would surprise me.)

It's already double damage but yeah why not

> > if you have a flyswatter and click on a tile that has someone with this mutation on it, regardless if you clicked on their sprite, it should crit them (or deal a shitload)
> 
> Insta gib fly people and instakill moff people with the tiny mutation? (Although I think being able to just click the same tile as them is pretty strong ontop of the double damage they will take.)
> 
> 
I mean combined with the double damage + increased flyswatter damage + moff flyswatter damage it would probably instakill them anyways

Megamaw commented:
frankly the thing that matters is 'make it less of a nightmare to click on them with the flyswatter if that's an issue'

LemonInTheDark commented:
The list of benefits here majorly outweighs the downsides, I don't want to end up with another vampire armory rush sort of situation. Could you explain the logic behind the different behaviors you're adding here?

c420-o commented:
> The list of benefits here majorly outweighs the downsides, I don't want to end up with another vampire armory rush sort of situation. Could you explain the logic behind the different behaviors you're adding here?

The goal is that you'd be easily abused (i.e. higher stamcrit damage = being able to be picked up and put in a crate where you'd be omega fucked, or nearly dying while naked venting). I'd suggest doing floyd's idea of test merging it. If there's any major unbalances with certain things I'd be glad to change them. As I said, I can't really do much with this atm since I have no way to reasonably test metabolism changes. If you plan on test merging it today I'd be able to add more negatives

Qustinnus commented:
i suggested a testmerge because i figured this code would break a ton of shit, not because I think the idea is worth trying out. It seems like an OP mutation.

c420-o commented:
> i suggested a testmerge because i figured this code would break a ton of shit, not because I think the idea is worth trying out. It seems like an OP mutation.

Yeah it probably won't get merged. Ah well

LemonInTheDark commented:
> 
> 
> > The list of benefits here majorly outweighs the downsides, I don't want to end up with another vampire armory rush sort of situation. Could you explain the logic behind the different behaviors you're adding here?
> 
> The goal is that you'd be easily abused (i.e. higher stamcrit damage = being able to be picked up and put in a crate where you'd be omega fucked, or nearly dying while naked venting). I'd suggest doing floyd's idea of test merging it. If there's any major unbalances with certain things I'd be glad to change them. As I said, I can't really do much with this atm since I have no way to reasonably test metabolism changes. If you plan on test merging it today I'd be able to add more negatives

I realize I've left you somewhat high and dry here, but I do think this is a bit much. It has soul, but I doubt anything could outweigh ventcrawling or tiny hitboxes. Not to mention the other issues with genetics and how people play it. Again, sorry for the wait here, shoulda said something a while ago

c420-o commented:
> > 
> > 
> > > The list of benefits here majorly outweighs the downsides, I don't want to end up with another vampire armory rush sort of situation. Could you explain the logic behind the different behaviors you're adding here?
> > 
> > The goal is that you'd be easily abused (i.e. higher stamcrit damage = being able to be picked up and put in a crate where you'd be omega fucked, or nearly dying while naked venting). I'd suggest doing floyd's idea of test merging it. If there's any major unbalances with certain things I'd be glad to change them. As I said, I can't really do much with this atm since I have no way to reasonably test metabolism changes. If you plan on test merging it today I'd be able to add more negatives
> 
> I realize I've left you somewhat high and dry here, but I do think this is a bit much. It has soul, but I doubt anything could outweigh ventcrawling or tiny hitboxes. Not to mention the other issues with genetics and how people play it. Again, sorry for the wait here, shoulda said something a while ago

It is what it is. I got some other cool pull requests you should merge tho :^)
